mono/metadata/superluminal.c

```diff
@@ -0,0 +1,324 @@
+	char *full_class_name = g_strdup_printf ("%s.%s", mono_class_get_name (klass), mono_method_get_name (method));
+	const char *namespace = mono_class_get_namespace (klass);
+	gpointer code_start = mono_jit_info_get_code_start (jinfo);
+	int code_size = mono_jit_info_get_code_size (jinfo);
+	MonoImage *image = mono_class_get_image (klass);
+	uint32_t method_token = mono_unity_method_get_token (method);
+
+	gunichar2 *namespace_utf16 = u8to16 (namespace);
+	gunichar2 *full_class_name_utf16 = u8to16 (full_class_name);
+	gunichar2 *signature_utf16 = u8to16 (signature);
+
+	if (is_rundown) {
+		EventWriteMethodDCEndVerbose_V2 ((uint64_t)method, (uint64_t)image, (uint64_t)code_start, code_size, method_token, 0x4, namespace_utf16, full_class_name_utf16, signature_utf16, 0, 0);
+		EventWriteMethodDCEndILToNativeMap ((uint64_t)method, 0, 0, compressed_num_lines, il_offsets, native_offsets, 0);
+	} else {
+		EventWriteMethodLoadVerbose_V2 ((uint64_t)method, (uint64_t)image, (uint64_t)code_start, code_size, method_token, 0x4, namespace_utf16, full_class_name_utf16, signature_utf16, 0, 0);
+		EventWriteMethodILToNativeMap ((uint64_t)method, 0, 0, compressed_num_lines, il_offsets, native_offsets, 0);
+	}
+
+	g_free (signature_utf16);
+	g_free (full_class_name_utf16);
+	g_free (namespace_utf16);
+	g_free (sourceFilePath);
+	g_free (full_class_name);
+}
+
+struct JITEnumerationData {
+	int mNumDomains;
+	int mNumAssemblies;
+	int mNumMethods;
+};
+
+static void
+method_jit_done (MonoProfiler *prof, MonoMethod *method, MonoJitInfo *jinfo)
+{
+	method_load (mono_domain_get (), method, jinfo, FALSE);
+}
+
+static void
+on_enumerate_assembly (MonoAssembly *assembly, void *user_data)
+{
+	struct JITEnumerationData *enumerationData = (struct JITEnumerationData *)user_data;
+	enumerationData->mNumAssemblies++;
+
+	MonoImage *image = mono_assembly_get_image (assembly);
+	image_event (image, TRUE, FALSE);
+}
+
+static void
+on_enumerate_jit_method (MonoDomain *domain, MonoMethod *method, MonoJitInfo *jinfo, void *user_data)
+{
+	struct JITEnumerationData *enumerationData = (struct JITEnumerationData *)user_data;
+	enumerationData->mNumMethods++;
+
+	method_load (domain, method, jinfo, TRUE);
+}
+
+static void
+on_enumerate_domain (MonoDomain *domain, void *user_data)
+{
+	struct JITEnumerationData *enumerationData = (struct JITEnumerationData *)user_data;
+	enumerationData->mNumDomains++;
+
+	// Iterate through each assembly
+	mono_domain_assembly_foreach (domain, on_enumerate_assembly, enumerationData);
+
+	// Iterate through each JIT'ed method
+	mono_domain_jit_foreach (domain, on_enumerate_jit_method, enumerationData);
+}
+
+static void
+on_attach ()
+{
+	SUPERLUMINAL_LOG ("Enumerating JIT data...");
+
+	struct JITEnumerationData enumerationData;
+	memset (&enumerationData, 0, sizeof (enumerationData));
+
+	// Iterate through each domain
+	mono_domain_foreach (on_enumerate_domain, &enumerationData);
```

Comment:
Does this need locking?

Reply:
Oh , maybe it does. Should we take the loader lock while iterating the domains?

@rovarma: Do you recall checking on this in the initial implementation of this profiler for Mono?

Reply:
Thanks @rovarma!

@joncham @UnityAlex: Do you know about the domain deleted/destroyed question?